File: web/static/js/structures/doubly.js
/**
 * doubly.js — Lógica de simulação da Lista Encadeada Dupla.
 *
 * Diferenças em relação à singly.js:
 * - type: 'doubly' no snapshot (renderer desenha setas prev)
 * - pointers inclui TAIL
 * - insertEnd é O(1) (TAIL conhecido)
 * - passos descrevem atualização de AMBOS os ponteiros (next e prev)
 */

function initStructurePage() {
  let nodes  = [];
  let nextId = 0;

  // ── UI ────────────────────────────────────────────────────────────────
  const { btnGenerate, btnExecute, inputSize, inputValue, inputIndex, inputNewVal } =
    StructureUI.bootstrap({ fallbackName: 'Lista Dupla', memoryType: 'doubly' });

  // ── Memory ────────────────────────────────────────────────────────────
  const _mem = MemoryHelpers.forLinked({
    type: 'doubly', bytesPerNode: 24, baseAddr: 0x3000,
    hashMult: 2246822519, rangeSize: 0x4000, alignMask: ~0x17,
  });

  function _buildMemory(nodeList, accessedNodeId) { return _mem.buildMemory(nodeList, accessedNodeId); }

  // ── Generate ─────────────────────────────────────────────────────────

  btnGenerate.addEventListener('click', () => {
    const size = Math.min(30, Math.max(2, parseInt(inputSize.value) || 5));
    nodes = Array.from({ length: size }, () => ({ id: nextId++, value: Math.floor(Math.random() * 90) + 1 }));
    _mem.reset();
    Animator.load([{
      description: 'Lista dupla gerada com valores aleatórios.',
      snapshot: _snapshot(nodes, []),
      memory:   _buildMemory(nodes, null),
    }]);
  });
  btnGenerate.click();

  // ── Execute ─────────────────────────────────────────────────────────

  btnExecute.addEventListener('click', () => {
    const op     = selectOp.value;
    const value  = parseInt(inputValue.value);
    const index  = parseInt(inputIndex.value);
    const newVal = parseInt(inputNewVal.value);
    let steps    = [];
    _mem.reset();

    switch (op) {
      case 'insertBegin':   steps = _insertBegin(value); break;
      case 'insertEnd':     steps = _insertEnd(value); break;
      case 'insertAt':      steps = _insertAt(index, value); break;
      case 'removeBegin':   steps = _removeBegin(); break;
      case 'removeEnd':     steps = _removeEnd(); break;
      case 'removeAt':      steps = _removeAt(index); break;
      case 'removeByValue': steps = _removeByValue(value); break;
      case 'search':        steps = _search(value); break;
      case 'update':        steps = _update(index, newVal); break;
    }

    if (steps.length > 0) {
      const lastSnap = steps[steps.length - 1].snapshot;
      nodes = lastSnap.nodes.map(n => ({ id: n.id, value: n.value }));
    }

    Animator.load(steps);
  });

  // ── Operations ────────────────────────────────────────────────────────

  function _insertBegin(val) {
    if (isNaN(val)) { alert('Informe um valor.'); return []; }
    const list = nodes.slice();
    const n    = { id: nextId++, value: val };
    const steps = [];

    steps.push({
      description: `Criar novo nó com valor ${val}.`,
      snapshot: _snapshot(list, [], 'neutral', [n], 'success'),
      memory:   _buildMemory([n, ...list], n.id),
    });

    steps.push({
      description: `Novo nó.next → antigo HEAD (${list[0]?.value ?? 'null'}); antigo HEAD.prev → novo nó.`,
      snapshot: _snapshot(list, list.length > 0 ? [0] : [], 'visiting', [n], 'success'),
      memory:   _buildMemory([n, ...list], list[0]?.id ?? null),
    });

    const result = [n, ...list];
    steps.push({
      description: `HEAD atualizado para o novo nó. Inserção no início concluída — O(1).`,
      snapshot: _snapshot(result, [0], 'success'),
      memory:   _buildMemory(result, null),
    });

    return steps;
  }

  function _insertEnd(val) {
    if (isNaN(val)) { alert('Informe um valor.'); return []; }
    const list = nodes.slice();
    const n    = { id: nextId++, value: val };
    const steps = [];

    steps.push({
      description: `Inserir ${val} no fim. Com TAIL disponível, acesso direto em O(1) — sem percorrer a lista.`,
      snapshot: _snapshot(list, []),
      memory:   _buildMemory(list, null),
    });

    if (list.length > 0) {
      steps.push({
        description: `TAIL aponta para o nó ${list[list.length - 1].value}. Novo nó.prev → TAIL; TAIL.next → novo nó.`,
        snapshot: _snapshot(list, [list.length - 1], 'warning', [n], 'success'),
        memory:   _buildMemory([...list, n], list[list.length - 1].id),
      });
    }

    const result = [...list, n];
    steps.push({
      description: `TAIL atualizado para o novo nó. Inserção no fim concluída — O(1) graças ao ponteiro TAIL.`,
      snapshot: _snapshot(result, [result.length - 1], 'success'),
      memory:   _buildMemory(result, null),
    });

    return steps;
  }

  function _insertAt(idx, val) {
    if (isNaN(val))  { alert('Informe um valor.'); return []; }
    if (isNaN(idx) || idx < 0 || idx > nodes.length) { alert('Índice inválido.'); return []; }
    if (idx === 0) return _insertBegin(val);
    if (idx === nodes.length) return _insertEnd(val);

    const list = nodes.slice();
    const n    = { id: nextId++, value: val };
    const steps = [];

    steps.push({
      description: `Inserir ${val} na posição ${idx}. Percorrendo...`,
      snapshot: _snapshot(list, []),
      memory:   _buildMemory(list, null),
    });

    for (let i = 0; i <= idx; i++) {
      steps.push({
        description: `Passo ${i + 1}: Nó ${i} (valor ${list[i]?.value}).${i === idx - 1 ? ' Predecessor encontrado.' : i === idx ? ' Nó que será deslocado.' : ''}`,
        snapshot: _snapshot(list, [i], i === idx - 1 ? 'warning' : 'visiting'),
        memory:   _buildMemory(list, list[i].id),
      });
    }

    steps.push({
      description: `Novo nó.next → nó ${idx}; novo nó.prev → nó ${idx - 1}. Atualizar ponteiros dos vizinhos.`,
      snapshot: _snapshot(list, [idx - 1, idx], 'warning', [n], 'success'),
      memory:   _buildMemory(list, list[idx - 1].id),
    });

    const result = [...list.slice(0, idx), n, ...list.slice(idx)];
    steps.push({
      description: `Nó ${idx - 1}.next e nó ${idx + 1}.prev atualizados. Inserção concluída.`,
      snapshot: _snapshot(result, [idx], 'success'),
      memory:   _buildMemory(result, null),
    });

    return steps;
  }

  function _removeBegin() {
    if (nodes.length === 0) { alert('Lista vazia.'); return []; }
    const list = nodes.slice();
    const steps = [];

    steps.push({
      description: `Remover HEAD (valor: ${list[0].value}). HEAD.next.prev será definido como null.`,
      snapshot: _snapshot(list, [0], 'danger'),
      memory:   _buildMemory(list, list[0].id),
    });

    if (list.length > 1) {
      steps.push({
        description: `Novo HEAD será o nó ${list[1].value}. Seu ponteiro prev definido como null.`,
        snapshot: _snapshot(list, [0, 1], 'warning'),
        memory:   _buildMemory(list, list[1].id),
      });
    }

    const result = list.slice(1);
    steps.push({
      description: `HEAD atualizado. Nó ${list[0].value} removido — O(1).`,
      snapshot: _snapshot(result, result.length > 0 ? [0] : [], 'success'),
      memory:   _buildMemory(result, null),
    });

    return steps;
  }

  function _removeEnd() {
    if (nodes.length === 0) { alert('Lista vazia.'); return []; }
    const list = nodes.slice();
    const steps = [];

    steps.push({
      description: `Remover TAIL (valor: ${list[list.length - 1].value}). Com TAIL: acesso direto em O(1).`,
      snapshot: _snapshot(list, [list.length - 1], 'danger'),
      memory:   _buildMemory(list, list[list.length - 1].id),
    });

    if (list.length > 1) {
      steps.push({
        description: `Novo TAIL será o nó ${list[list.length - 2].value}. Seu ponteiro next definido como null.`,
        snapshot: _snapshot(list, [list.length - 2, list.length - 1], 'warning'),
        memory:   _buildMemory(list, list[list.length - 2].id),
      });
    }

    const result = list.slice(0, -1);
    steps.push({
      description: `TAIL atualizado. Nó ${list[list.length - 1].value} removido — O(1) graças ao TAIL.`,
      snapshot: _snapshot(result, []),
      memory:   _buildMemory(result, null),
    });

    return steps;
  }

  function _removeAt(idx) {
    if (isNaN(idx) || idx < 0 || idx >= nodes.length) { alert('Índice inválido.'); return []; }
    if (idx === 0) return _removeBegin();
    if (idx === nodes.length - 1) return _removeEnd();

    const list = nodes.slice();
    const steps = [];

    steps.push({
      description: `Remover nó na posição ${idx}. Percorrendo...`,
      snapshot: _snapshot(list, []),
      memory:   _buildMemory(list, null),
    });

    for (let i = 0; i <= idx; i++) {
      steps.push({
        description: `Passo ${i + 1}: Visitando nó ${i} (valor ${list[i].value}).${i === idx ? ' Nó a remover.' : ''}`,
        snapshot: _snapshot(list, [i], i === idx ? 'danger' : 'visiting'),
        memory:   _buildMemory(list, list[i].id),
      });
    }

    steps.push({
      description: `Atualizando: nó ${idx - 1}.next → nó ${idx + 1}; nó ${idx + 1}.prev → nó ${idx - 1}.`,
      snapshot: _snapshot(list, [idx - 1, idx, idx + 1], 'warning'),
      memory:   _buildMemory(list, list[idx - 1].id),
    });

    const result = [...list.slice(0, idx), ...list.slice(idx + 1)];
    steps.push({
      description: `Nó ${list[idx].value} removido. Ambos os ponteiros atualizados — lista permanece bidirecional.`,
      snapshot: _snapshot(result, []),
      memory:   _buildMemory(result, null),
    });

    return steps;
  }

  function _removeByValue(val) {
    if (isNaN(val)) { alert('Informe um valor.'); return []; }
    const list  = nodes.slice();
    const steps = [];

    steps.push({
      description: `Buscar valor ${val} para remover...`,
      snapshot: _snapshot(list, []),
      memory:   _buildMemory(list, null),
    });

    const idx = list.findIndex(n => n.value === val);
    if (idx === -1) {
      steps.push({
        description: `Valor ${val} não encontrado.`,
        snapshot: _snapshot(list, []),
        memory:   _buildMemory(list, null),
      });
      return steps;
    }

    for (let i = 0; i <= idx; i++) {
      const found = i === idx;
      steps.push({
        description: `Passo ${i + 1}: Nó ${i} (valor ${list[i].value}). ${found ? 'ENCONTRADO!' : 'Avançar.'}`,
        snapshot: _snapshot(list, [i], found ? 'danger' : 'visiting'),
        memory:   _buildMemory(list, list[i].id),
      });
    }

    return steps.concat(_removeAt(idx).slice(1));
  }

  function _search(val) {
    if (isNaN(val)) { alert('Informe um valor.'); return []; }
    const list  = nodes.slice();
    const steps = [];

    steps.push({
      description: `Busca linear pelo valor ${val} a partir do HEAD...`,
      snapshot: _snapshot(list, []),
      memory:   _buildMemory(list, null),
    });

    for (let i = 0; i < list.length; i++) {
      const found = list[i].value === val;
      steps.push({
        description: `Passo ${i + 1}/${list.length}: Nó ${i} tem valor ${list[i].value}. ${found ? 'ENCONTRADO!' : 'Avançar via next.'}`,
        snapshot: _snapshot(list, [i], found ? 'success' : 'visiting'),
        memory:   _buildMemory(list, list[i].id),
      });
      if (found) return steps;
    }

    steps.push({
      description: `Valor ${val} não encontrado.`,
      snapshot: _snapshot(list, []),
      memory:   _buildMemory(list, null),
    });
    return steps;
  }

  function _update(idx, val) {
    if (isNaN(idx) || idx < 0 || idx >= nodes.length) { alert('Índice inválido.'); return []; }
    if (isNaN(val)) { alert('Informe o novo valor.'); return []; }

    const list  = nodes.slice();
    const steps = [];

    steps.push({
      description: `Atualizar nó na posição ${idx}. Percorrendo...`,
      snapshot: _snapshot(list, []),
      memory:   _buildMemory(list, null),
    });

    for (let i = 0; i <= idx; i++) {
      steps.push({
        description: `Passo ${i + 1}: Nó ${i} (valor ${list[i].value}).${i === idx ? ' Posição encontrada.' : ''}`,
        snapshot: _snapshot(list, [i], i === idx ? 'warning' : 'visiting'),
        memory:   _buildMemory(list, list[i].id),
      });
    }

    const result = list.map((n, i) => i === idx ? { ...n, value: val } : n);
    steps.push({
      description: `Nó ${idx} atualizado: ${list[idx].value} → ${val}. Ponteiros não precisam ser alterados.`,
      snapshot: _snapshot(result, [idx], 'success'),
      memory:   _buildMemory(result, null),
    });

    return steps;
  }

  // ── Snapshot ───────────────────────────────────────────────────────────

  function _snapshot(list, highlightedIdxs = [], hlState = 'visiting', prependNodes = [], prependState = 'neutral') {
    const allNodes = [
      ...prependNodes.map(n => ({ id: n.id, value: n.value, state: prependState })),
      ...list.map((n, i) => ({
        id:    n.id,
        value: n.value,
        state: highlightedIdxs.includes(i) ? hlState : 'neutral',
      })),
    ];
    const headId = allNodes[0]?.id;
    const tailId = allNodes[allNodes.length - 1]?.id;
    return { type: 'doubly', nodes: allNodes, pointers: { HEAD: headId, TAIL: tailId } };
  }

}

File: internal/structures/linkedlist/doubly.go
package linkedlist

import "errors"

// nodeD é um nó da lista duplamente encadeada.
// Cada nó possui ponteiros para o próximo (Next) E para o anterior (Prev).
type nodeD struct {
	Value int
	Next  *nodeD
	Prev  *nodeD
}

// DoublyLinkedList possui HEAD (início) e TAIL (fim), permitindo
// percorrer a lista nos dois sentidos e inserção/remoção no fim em O(1).
type DoublyLinkedList struct {
	Head *nodeD
	Tail *nodeD
	size int
}

// InsertBegin insere no início — O(1).
// O novo nó se torna HEAD; seu Next aponta para o antigo HEAD;
// o antigo HEAD passa a ter Prev apontando para o novo nó.
func (l *DoublyLinkedList) InsertBegin(value int) {
	n := &nodeD{Value: value}
	if l.Head == nil {
		l.Head = n
		l.Tail = n
	} else {
		n.Next = l.Head   // novo nó aponta para o antigo HEAD
		l.Head.Prev = n   // antigo HEAD passa a ter predecessor
		l.Head = n        // HEAD atualizado
	}
	l.size++
}

// InsertEnd insere no final usando TAIL — O(1) (vantagem sobre a lista simples).
func (l *DoublyLinkedList) InsertEnd(value int) {
	n := &nodeD{Value: value}
	if l.Tail == nil {
		l.Head = n
		l.Tail = n
	} else {
		n.Prev = l.Tail   // novo nó aponta para o antigo TAIL
		l.Tail.Next = n   // antigo TAIL aponta para o novo nó
		l.Tail = n        // TAIL atualizado
	}
	l.size++
}

// InsertAt insere na posição index — O(n).
func (l *DoublyLinkedList) InsertAt(index, value int) error {
	if index < 0 || index > l.size {
		return errors.New("índice fora dos limites")
	}
	if index == 0 {
		l.InsertBegin(value)
		return nil
	}
	if index == l.size {
		l.InsertEnd(value)
		return nil
	}

	// Avança até o nó que atualmente ocupa a posição index.
	current := l.Head
	for i := 0; i < index; i++ {
		current = current.Next
	}

	n := &nodeD{Value: value}
	prev := current.Prev

	// Liga o novo nó entre prev e current.
	n.Prev = prev
	n.Next = current
	prev.Next = n
	current.Prev = n

	l.size++
	return nil
}

// RemoveBegin remove o primeiro nó — O(1).
func (l *DoublyLinkedList) RemoveBegin() (int, error) {
	if l.Head == nil {
		return 0, errors.New("lista vazia")
	}
	value := l.Head.Value
	l.Head = l.Head.Next
	if l.Head != nil {
		l.Head.Prev = nil // novo HEAD não tem predecessor
	} else {
		l.Tail = nil // lista ficou vazia
	}
	l.size--
	return value, nil
}

// RemoveEnd remove o último nó usando TAIL — O(1).
func (l *DoublyLinkedList) RemoveEnd() (int, error) {
	if l.Tail == nil {
		return 0, errors.New("lista vazia")
	}
	value := l.Tail.Value
	l.Tail = l.Tail.Prev
	if l.Tail != nil {
		l.Tail.Next = nil // novo TAIL não tem sucessor
	} else {
		l.Head = nil // lista ficou vazia
	}
	l.size--
	return value, nil
}

// RemoveAt remove o nó na posição index — O(n).
func (l *DoublyLinkedList) RemoveAt(index int) (int, error) {
	if index < 0 || index >= l.size {
		return 0, errors.New("índice fora dos limites")
	}
	if index == 0 {
		return l.RemoveBegin()
	}
	if index == l.size-1 {
		return l.RemoveEnd()
	}

	current := l.Head
	for i := 0; i < index; i++ {
		current = current.Next
	}

	// Reconecta os vizinhos, "pulando" o nó atual.
	current.Prev.Next = current.Next
	current.Next.Prev = current.Prev

	l.size--
	return current.Value, nil
}

// RemoveByValue remove a primeira ocorrência do valor — O(n).
func (l *DoublyLinkedList) RemoveByValue(value int) error {
	current := l.Head
	for current != nil {
		if current.Value == value {
			if current.Prev != nil {
				current.Prev.Next = current.Next
			} else {
				l.Head = current.Next
			}
			if current.Next != nil {
				current.Next.Prev = current.Prev
			} else {
				l.Tail = current.Prev
			}
			l.size--
			return nil
		}
		current = current.Next
	}
	return errors.New("valor não encontrado")
}

// Search retorna o índice da primeira ocorrência — O(n).
func (l *DoublyLinkedList) Search(value int) (int, bool) {
	current := l.Head
	for i := 0; current != nil; i++ {
		if current.Value == value {
			return i, true
		}
		current = current.Next
	}
	return -1, false
}

// Update substitui o valor do nó na posição index — O(n).
func (l *DoublyLinkedList) Update(index, value int) error {
	if index < 0 || index >= l.size {
		return errors.New("índice fora dos limites")
	}
	current := l.Head
	for i := 0; i < index; i++ {
		current = current.Next
	}
	current.Value = value
	return nil
}

// Size retorna o número de nós.
func (l *DoublyLinkedList) Size() int { return l.size }

// ToSlice retorna os valores em ordem.
func (l *DoublyLinkedList) ToSlice() []int {
	result := make([]int, 0, l.size)
	current := l.Head
	for current != nil {
		result = append(result, current.Value)
		current = current.Next
	}
	return result
}
